### ISSUE TITLE
PRIME Render Offload support

### ISSUE THREAD

**Issue by Zeioth** [Askannz/optimus-manager#114]
Nvidia finally supports PRIME Render Offload. Now it should be possible to switch GPU without having to restart the xorg session. More info [here](https://www.phoronix.com/scan.php?page=news_item&px=NVIDIA-435.17-Linux-Driver&fbclid=IwAR0YdVYS0BTGIs27UY_7U_FMdYixDkYsDkZnyer6Aig1ZqaRyONO-ox6Ggc).

**Comment by Askannz** [comment]
Yep, saw that :slightly_smiling_face: 

This is definitely good news, but as I see it now the solution offered by Nvidia has a few major limitations :

* You still cannot use video outputs attached to your Nvidia card when the X server is running on Intel (yet).

* The nvidia module has to stay loaded for offloading to work, which means that bbswitch or ACPI calls to turn off the GPU are not an option. They seem to want to rely on dynamic power management inside the driver instead, [but they've only implemented it for Turing cards and above](http://download.nvidia.com/XFree86/Linux-x86_64/435.17/README/dynamicpowermanagement.html). That says, even with older cards, the power consumption should still be lower than running the X server at full blast on the Nvidia card, because it can just idle in the background and clock to a low power state.

So I don't think it makes optimus-manager irrelevant just yet. I'll just add an option to keep the nvidia driver loaded even in Intel mode, so that the dynamic power management can kick in and the card stays available for offloading (it should allow for CUDA applications to run in Intel mode, too).

**Comment by openmindead** [comment]
I think it's priceless to have whole desktop running on Intel since their chips are powerful enough for this task, and being able to run demanding or special tasks on Nvidia at the same time. Running KDE, for instance, on some weak Nvidia GPUs is something that always drived me nuts. In such case there are only 2 reasonable scenarios:
1. Only Intel.
2. Intel with Nvidia offload.

**Comment by pssncp142** [comment]
> So I don't think it makes optimus-manager irrelevant just yet. I'll just add an option to keep the nvidia driver loaded even in Intel mode, so that the dynamic power management can kick in and the card stays available for offloading (it should allow for CUDA applications to run in Intel mode, too).

I think a new mod is better than adding an option into intel mode. Due to power management issues on older cards, I am still not convinced to run offloading option on my laptop all the time but it is much better than running X on nvidia card.

I have a new branch with quick edits if you would like to try offloading feature with `--switch hybrid`:

https://github.com/pssncp142/optimus-manager/tree/nvidia-hybrid

I can't guarantee that it will work for everyone. It is just a temporary solution on my end for testing.

**Comment by Askannz** [comment]
@pssncp142 yeah, you're right. It makes more sense to have this feature in a new mode because users may still want to have access to the full power switching.

Would you mind making a pull request with your changes ? I'll do some testing and edit if necessary.

**Comment by sebirdman** [comment]
@pssncp142 I tried your patch it seems to work well however my GPU does not seem to shut down (1650) and nvidia-smi still reports 3W of power use. 

I think it should, it's got the Turing architecture: https://www.nvidia.com/en-us/geforce/graphics-cards/gtx-1650/

**Comment by pssncp142** [comment]
@sebirdman Yes, 1650 should support it. However, my patch doesn't include dynamic power management. See here for detailed information:
http://download.nvidia.com/XFree86/Linux-x86_64/435.17/README/dynamicpowermanagement.html

I don't have hardware to test dynamic power management. My laptop has 1050. I applied the logic here: https://github.com/pssncp142/optimus-manager/tree/nvidia-hybrid-dpm, as far as I understand it. If you would like to test it, give it a go. Be sure to set `switching=none`.

**Comment by Zeioth** [comment]
@sebirdman it's pretty impressive that the GPU only consumes 3W on standby. I think it's worth it even for pre-turing users.

A laptop usually consumes 40–75W. (4W if the screen is turned off) so that's only about a 6% increase.

**Comment by Askannz** [comment]
@pssncp142's pull request was merged into master, by the way :slightly_smiling_face:

There is more stuff I want of add before releasing a new version, but feel free to test it.

@sebirdman maybe your CPU isn't recent enough ? According to http://download.nvidia.com/XFree86/Linux-x86_64/435.17/README/dynamicpowermanagement.html, the dynamic power management also requires a Coffee Lake CPU.

**Comment by sebirdman** [comment]
@Askannz @pssncp142 with the dynamic power stuff enabled, it looks like powertop says the card is off. When i run nvidia-smi i now see:

```
Sun Aug 25 19:20:16 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 435.17       Driver Version: 435.17       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1650    Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   42C    P8     1W /  N/A |     38MiB /  3911MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1354      G   /usr/lib/Xorg                                 36MiB |
+-----------------------------------------------------------------------------+
WARNING: infoROM is corrupted at gpu 0000:01:00.0
```

**Comment by Askannz** [comment]
Thanks, so power consumption looks good. However I'm a bit worried about that line :
```
WARNING: infoROM is corrupted at gpu 0000:01:00.0
```
It could be just a side effect of the power management, but I also made some changes to the way optimus-manager loads kernel modules, and I'm worried this may be the cause.

If you don't mind, could you test the hybrid mode with commit 348fdd10b069e05398a382378172ef255dca9304 and see if the same message pops up ?

**Comment by pssncp142** [comment]
Good to see power management working @sebirdman ! (ignoring the warning for the moment).

Quoting nvidia:

> Option "NVreg_DynamicPowerManagement=0x02"
> With this setting, the NVIDIA GPU driver will allow the GPU to go into its lowest power state when no applications are running that use the nvidia driver stack. Whenever an application requiring NVIDIA GPU access is started, the GPU is put into an active state. When the application exits, the GPU is put into a low power state.

So, I would expect `nvidia-smi` to wake up the GPU temporarily.

@Askannz Thanks for the good work and your help. I am leaving the rest to you since I can't use the power management feature.

**Comment by Askannz** [comment]
> It also looks like a poster on: https://devtalk.nvidia.com/default/topic/1061293/linux/-435-17-am-i-observing-the-proper-results-with-the-new-prime-/post/5375031/#5375031 also had that warning
Good, probably not my fault then :smile: 

> Also one thing that kinda throws me off is that `optimus-manager --print-mode` shows "intel" i'd almost expect it to print "hybrid" but it looks like print-mode gets its info from glxinfo.
Good point, I'll fix that.

**Comment by felagund** [comment]
> * You still cannot use video outputs attached to your Nvidia card when the X server is running on Intel (yet).

What exactly does that "yet" mean? That they might add it in the future or  that they are working on it?

**Comment by Algebro7** [comment]
@sebirdman can you provide instructions for getting the dynamic power management working? It wasn't clear to me from the Nvidia docs what exactly I needed to do to enable it on the 1650.

**Comment by sebirdman** [comment]
@Algebro7 i added a modprobe setting:

in /etc/modprobe.d/nvidia.conf 
```
options nvidia "NVreg_DynamicPowerManagement=0x02"
```

**Comment by x0wllaar** [comment]
@sebirdman How do you check if the card is off with powertop though?
```
$ nvidia-smi
Fri Oct 11 04:24:48 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 435.21       Driver Version: 435.21       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 2060    Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   45C    P8     4W /  N/A |     22MiB /  5934MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      8938      G   /usr/lib/xorg/Xorg                            20MiB |
+-----------------------------------------------------------------------------+
```

This is with the option set in /etc/modprobe.d/nvidia.conf.
I'm not sure if the card is actually off.

**Comment by sebirdman** [comment]
@x0wllaar i see nothing in dmesg when i run that

After bootup, it does take a second to spin down. 

```
$ nvidia-smi
Thu Oct 10 21:35:57 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 435.21       Driver Version: 435.21       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1650    Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   42C    P8     1W /  N/A |     38MiB /  3911MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     14742      G   /usr/lib/Xorg                                 36MiB |
+-----------------------------------------------------------------------------+
WARNING: infoROM is corrupted at gpu 0000:01:00.0
```

**Comment by x0wllaar** [comment]
@sebirdman, Thanks!

My GPU doesn't go to sleep at  all then. I wonder if this is because I have a Ryzen CPU, although it and it's SoC should support low power states. 

Maybe the driver doesn't try to put the GPU to sleep unless it detects an Intel CPU.

**Comment by Askannz** [comment]
@felagund 
> > * You still cannot use video outputs attached to your Nvidia card when the X server is running on Intel (yet).
> 
> What exactly does that "yet" mean? That they might add it in the future or that they are working on it?

Yeah, an Nvidia engineer confirmed that it's planned. But he didn't give any ETA, so who knows when the feature will actually land ¯\_(ツ)_/¯


@x0wllaar 
> @sebirdman, Thanks!
> 
> My GPU doesn't go to sleep at all then. I wonder if this is because I have a Ryzen CPU, although it and it's SoC should support low power states.
> 
> Maybe the driver doesn't try to put the GPU to sleep unless it detects an Intel CPU.

You need an Intel Coffee Lake CPU for the feature to work: http://download.nvidia.com/XFree86/Linux-x86_64/435.17/README/dynamicpowermanagement.html

> The necessary hardware and ACPI support was first added in Intel Coffeelake chipset series. Hence, this feature is supported from Intel Coffeelake chipset series.

**Comment by Askannz** [comment]
Support for `--hybrid` mode is implemented since v1.2, so I'm closing this. If you experience specific bugs with this feature, please report them in separate issues.
